scripts/scanner.lua

```diff
@@ -407,8 +436,8 @@ function mod.list_sort(pindex)
    for i, name in ipairs(players[pindex].nearby.ents   ) do
       local i1 = 1
       while i1 <= #name.ents do --this appears to be removing invalid ents within a set.
-         if name.ents[i1] == nil or (name.ents[i1].valid == false and name.aggregate == false) then
-            table.remove(name.ents, i1)
+         if name.ents[i1].valid == false and name.aggregate == false then
```

Comment:
Factorio shouldn't be changing entities out from under us in the same event handler---maybe you know something I don't but I'd be astonished.  But even then it doesn't go to nil, just to invalid.  The reason I dislike nils in Lua specifically but even more generally in other languages is that there is usually a way to not need them, and they add this sort of complexity--either you check too much, check too little, or spend time having this exact discussion and having to think deeply.

But in this case it's simpler than that and we can avoid digging super deeply into this.  From the manual:

```
Unless a __len metamethod is given, the length of a table t is only defined if the table is a sequence, that is, the set of its positive numeric keys is equal to {1..n} for some non-negative integer n. In that case, n is its length.
```

So if there are "holes" we've got bigger problems.  You get a length but it's not actually guaranteed to be the right length.  I did some digging and at least PUC Lua is using some sort of binary search, so you may or may not see the holes and that will vary based on where the holes happen to be.  Factorio probably makes this deterministic, but they definitely don't document what it is.  ?  Worst case this moves a bug we almost certainly already have to a different bug, and for that matter the different bug will trigger an error with stack traces rather than silently being weird.

But in general stuff like this is why I said I want to bake it for a while, and why I reached out to Zack to run from source.

Reply:
I think in this case the surface find function returns an empty table rather than nil, but we still attempt to reference element number 1, which is why it could be nil.